ocaml/libs/clock/date.ml

```diff
@@ -0,0 +1,169 @@
+(* Copyright (C) Cloud Software Group Inc.
+   This program is free software; you can redistribute it and/or modify
+   it under the terms of the GNU Lesser General Public License as published
+   by the Free Software Foundation; version 2.1 only. with the special
+   exception on linking described in file LICENSE.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU Lesser General Public License for more details.
+*)
+
+let months =
+  [|
+     "Jan"
+   ; "Feb"
+   ; "Mar"
+   ; "Apr"
+   ; "May"
+   ; "Jun"
+   ; "Jul"
+   ; "Aug"
+   ; "Sep"
+   ; "Oct"
+   ; "Nov"
+   ; "Dec"
+  |]
+
+let days = [|"Sun"; "Mon"; "Tue"; "Wed"; "Thu"; "Fri"; "Sat"|]
+
+type print_timezone = Empty | TZ of string
+
+(* we must store the print_type with iso8601 to handle the case where the local time zone is UTC *)
+type t = Ptime.date * Ptime.time * print_timezone
+
+let utc = TZ "Z"
+
+let of_dt print_type dt =
+  let date, time = dt in
+  (date, time, print_type)
+
+let to_dt (date, time, _) = (date, time)
+
+let best_effort_iso8601_to_rfc3339 x =
+  (* (a) add dashes
+   * (b) add UTC tz if no tz provided *)
+  let x =
+    try
+      Scanf.sscanf x "%04d%02d%02dT%s" (fun y mon d rest ->
+          Printf.sprintf "%04d-%02d-%02dT%s" y mon d rest
+      )
+    with _ -> x
+  in
+  let tz =
+    try
+      Scanf.sscanf x "%04d-%02d-%02dT%02d:%02d:%02d%s" (fun _ _ _ _ _ _ tz ->
+          Some tz
+      )
+    with _ -> None
+  in
+  match tz with
+  | None | Some "" ->
+      (* the caller didn't specify a tz. we must try to add one so that ptime can at least attempt to parse *)
+      (Printf.sprintf "%sZ" x, Empty)
+  | Some tz ->
+      (x, TZ tz)
+
+let of_iso8601 x =
+  let rfc3339, print_timezone = best_effort_iso8601_to_rfc3339 x in
+  match Ptime.of_rfc3339 rfc3339 |> Ptime.rfc3339_error_to_msg with
+  | Error _ ->
+      invalid_arg (Printf.sprintf "%s: %s" __FUNCTION__ x)
+  | Ok (t, tz, _) -> (
+    match tz with
+    | None | Some 0 ->
+        Ptime.to_date_time t |> of_dt print_timezone
+    | Some _ ->
+        invalid_arg (Printf.sprintf "%s: %s" __FUNCTION__ x)
+  )
+
+let to_rfc3339 ((y, mon, d), ((h, min, s), _), print_type) =
+  match print_type with
+  | TZ tz ->
+      Printf.sprintf "%04i%02i%02iT%02i:%02i:%02i%s" y mon d h min s tz
+  | Empty ->
+      Printf.sprintf "%04i%02i%02iT%02i:%02i:%02i" y mon d h min s
+
+let weekday ~year ~mon ~day =
```

Comment:
A reference to this algorithm (as a comment) would be helpful because it's not evident that this is correct.

Reply:
I can add it, the reference is https://www.tondering.dk/claus/cal/chrweek.php#calcdow

Reply:
The function is tested as part of the email format generator, I picked a couple of dates in the past and compared it with the weekday shown by newspapers emitted that day